### PR TITLE
ARROW-15593: [C++] Fix a race condition in the thread pool

### LINKED ISSUE #12329: An unhandled race condition exists in `ThreadPool`
As the title mentioned, I suspect that an unhandled race condition exists here: https://github.com/apache/arrow/blob/f48cabee64fe0719e5e6f28d89ca86277373ca2b/cpp/src/arrow/util/thread_pool.cc#L238-L257

Assume that we have two threads, one executes `SpawnReal` to push the task into the `pending_tasks`, the other is running at the `ProtectAgainstFork`, which may reset the whole `state` which the former one is using, which makes the loss of tasks. 

In my scenario, when I fork multiple processes to read data after reading something in the parent process, it has chances to stall at `Future::Wait`.  (However, I'm not very confident that the race condition is the root cause of my scenario since I only reproduce it in python code, but fail to reproduce it in cpp.)

A quick fix could be change the type of `pid_` to `std::atomic<pid_t>` and  use `compare_exchange_strong` to protect the critical section.

### PR DESCRIPTION
**Pull request by framlog**
Test Plan: Manually

This closes #12329.

### CODE REVIEW

File: cpp/src/arrow/util/thread_pool.h [hunk 1]
```diff
@@ -373,7 +374,7 @@ class ARROW_EXPORT ThreadPool : public Executor {
   State* state_;
   bool shutdown_on_destroy_;
 #ifndef _WIN32
-  pid_t pid_;
+  std::atomic<pid_t> pid_;
```
Comment (cyb70289) [inline cpp/src/arrow/util/thread_pool.h:377]: This gives an uninitialized random value.
Zero initialize with `std::atomic<pid_t> pid_{}`.

Comment (framlog) [reply cpp/src/arrow/util/thread_pool.h:377]: It's initialized at https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/thread_pool.cc#L225. And I don't think it should be zero when it's initialized.

File: cpp/src/arrow/util/thread_pool.cc [hunk 2]
```diff
@@ -235,23 +235,32 @@ ThreadPool::~ThreadPool() {
...
-
-    pid_ = current_pid;
-    sp_state_ = new_state;
-    state_ = sp_state_.get();
-
+  pid_t old_pid = pid_.load();
+  int capacity = 0;
+  if (current_pid != old_pid) {
+    // FIXME: It may get deadlock here if users fork processes while the parent runs
+    // in the middle of resetting states.
```
Comment (pitrou) [inline cpp/src/arrow/util/thread_pool.cc:242]: FTR, I'm working on a better fix for this.

**Comment by pitrou** [comment]
See PR #12358. The tests I added there sometimes fail with this PR.

File: cpp/src/arrow/util/thread_pool.cc (hunk 2 above)
Comment (framlog) [reply cpp/src/arrow/util/thread_pool.cc:242]: Great. Feel free to close this then.

**Comment by cyb70289** [comment]
Close this PR as it's fixed in #12358.
Thanks @framlog to investigate this tricky issue !

**Pull request closed without merging**